File: src/main/java/org/gradle/plugins/site/generator/FreemarkerSiteGenerator.java
package org.gradle.plugins.site.generator;

import freemarker.template.Configuration;
import freemarker.template.Template;
import freemarker.template.TemplateException;
import freemarker.template.TemplateExceptionHandler;
import org.gradle.api.GradleException;
import org.gradle.plugins.site.data.CustomData;
import org.gradle.plugins.site.data.ProjectDescriptor;
import org.gradle.plugins.site.utils.FileUtils;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * A site generator implementation based on <a href="http://freemarker.org/">Freemarker</a>.
 */
public class FreemarkerSiteGenerator implements SiteGenerator {

    private final File outputDir;

    public FreemarkerSiteGenerator(File outputDir) {
        this.outputDir = outputDir;
    }

    public void generate(ProjectDescriptor projectDescriptor, CustomData customData) {
        try {
            copyCssResources();
            copyImgResources();
            processIndexPageTemplate(projectDescriptor, customData);
        } catch (Exception e) {
            throw new GradleException("Unable to generate site", e);
        }
    }

    private void copyCssResources() throws IOException, URISyntaxException {
        List<String> resources = new ArrayList<String>();
        resources.add("bootstrap.css");
        resources.add("bootstrap-responsive.css");
        copyResources("css", resources);
    }

    private void copyImgResources() throws IOException, URISyntaxException {
        List<String> resources = new ArrayList<String>();
        resources.add("elephant-corner.png");
        copyResources("img", resources);
    }

    private void copyResources(String subdir, List<String> resources) throws IOException, URISyntaxException {
        File targetDir = new File(outputDir, subdir);
        FileUtils.createDirectory(targetDir);

        for (String resource : resources) {
            String sourcePath = subdir + "/" + resource;
            FileUtils.copyFile(resolveAsUrl(sourcePath), new File(targetDir, resource));
        }
    }

    private InputStream resolveAsUrl(String name) {
        return getClass().getClassLoader().getResourceAsStream(name);
    }

    private void processIndexPageTemplate(ProjectDescriptor projectDescriptor, CustomData customData) throws IOException, URISyntaxException, TemplateException {
        Configuration cfg = new Configuration(Configuration.VERSION_2_3_25);
        cfg.setClassLoaderForTemplateLoading(getClass().getClassLoader(), "template");
        cfg.setDefaultEncoding("UTF-8");
        cfg.setTemplateExceptionHandler(TemplateExceptionHandler.RETHROW_HANDLER);
        cfg.setLogTemplateExceptions(false);
        Map<String, Object> root = new HashMap<String, Object>();
        root.put("project", projectDescriptor);
        root.put("customData", customData);
        Template template = cfg.getTemplate("index.ftl");
        template.process(root, new FileWriter(new File(outputDir, "index.html")));
    }
}

File: src/main/kotlin/org/gradle/plugins/site/generator/FreemarkerSiteGenerator.kt
package org.gradle.plugins.site.generator

import freemarker.template.Configuration
import freemarker.template.TemplateException
import freemarker.template.TemplateExceptionHandler
import org.gradle.api.GradleException
import org.gradle.plugins.site.data.ProjectDescriptor
import org.gradle.plugins.site.data.ProjectLinksDescriptor
import java.io.File
import java.io.IOException
import java.io.InputStream

/**
 * A site generator implementation based on [Freemarker](https://freemarker.org/).
 */
class FreemarkerSiteGenerator(private val outputDir: File) : SiteGenerator {

    override fun generate(projectDescriptor: ProjectDescriptor, projectLinksDescriptor: ProjectLinksDescriptor) {
        try {
            copyCssResources()
            copyImgResources()
            processIndexPageTemplate(projectDescriptor, projectLinksDescriptor)
        } catch (e: Exception) {
            throw GradleException("Unable to generate site", e)
        }

    }

    @Throws(IOException::class)
    private fun copyCssResources() {
        val resources = listOf(
                "bootstrap.css",
                "bootstrap-responsive.css"
        )
        copyResources("css", resources)
    }

    @Throws(IOException::class)
    private fun copyImgResources() {
        val resources = listOf(
                "elephant-corner.png"
        )
        copyResources("img", resources)
    }

    @Throws(IOException::class)
    private fun copyResources(subdir: String, resources: List<String>) {
        val targetDir = outputDir.resolve(subdir)
        for (resource in resources) {
            resolveAsStream("$subdir/$resource").use { inputStream ->
                inputStream.copyToFile(targetDir.resolve(resource))
            }
        }
    }

    private fun resolveAsStream(name: String): InputStream {
        return javaClass.classLoader.getResourceAsStream(name).buffered()
    }

    @Throws(IOException::class)
    private fun InputStream.copyToFile(file: File) {
        file.parentFile.let { dir ->
            if (!dir.isDirectory && !dir.mkdirs()) throw IOException("Unable to create directory $dir")
        }
        file.outputStream().buffered().use { outputStream ->
            copyTo(outputStream)
        }
    }

    @Throws(IOException::class, TemplateException::class)
    private fun processIndexPageTemplate(projectDescriptor: ProjectDescriptor, linksDescriptor: ProjectLinksDescriptor) {
        val cfg = Configuration(Configuration.VERSION_2_3_25).apply {
            setClassLoaderForTemplateLoading(javaClass.classLoader, "template")
            defaultEncoding = "UTF-8"
            templateExceptionHandler = TemplateExceptionHandler.RETHROW_HANDLER
            logTemplateExceptions = false
        }
        val root = mapOf(
                "project" to projectDescriptor,
                "customData" to linksDescriptor
        )
        val template = cfg.getTemplate("index.ftl")
        outputDir.resolve("index.html").bufferedWriter().use { writer ->
            template.process(root, writer)
        }
    }
}
